RedditSharp/WebAgent.cs

```diff
@@ -112,13 +66,38 @@ public int RequestsThisBurst
         public WebAgent( string accessToken ) {
             RootDomain = OAuthDomainUrl;
             AccessToken = accessToken;
+            RateLimitUsed = 0;
+            RateLimitReset = DateTimeOffset.UtcNow;
+            RateLimitRemaining = IsOAuth ? 60 : 30;
         }
 
         /// <inheritdoc />
-        public virtual async Task<JToken> ExecuteRequestAsync(HttpRequestMessage request)
+        public virtual async Task<JToken> ExecuteRequestAsync(Func<HttpRequestMessage> request)
         {
-            await EnforceRateLimit().ConfigureAwait(false);
-            var response = await _httpClient.SendAsync(request).ConfigureAwait(false);
+            if (request == null)
+              throw new ArgumentNullException(nameof(request));
+            const int maxTries = 20;
+            HttpResponseMessage response;
+            var tries = 0;
+            do {
+              await rateLimitLock.WaitAsync().ConfigureAwait(false);
```

Comment:
I was just thinking about it and realized that we could avoid race conditions by only setting "remaining" amount if it is lower than the current value or it reset.

How do you suggest handling it as a fallback?